02_Util/src/security/SignedMeterValuesUtil.ts

```diff
@@ -0,0 +1,209 @@
+import {
+  IChargingStationSecurityInfoRepository,
+  IDeviceModelRepository,
+  sequelize,
+} from '@citrineos/data';
+import {
+  IFileAccess,
+  MeterValueType,
+  SignedMeterValuesConfig,
+  SignedMeterValueType,
+  SystemConfig,
+} from '@citrineos/base';
+import { ILogObj, Logger } from 'tslog';
+import * as crypto from 'node:crypto';
+
+/**
+ * Util to process and validate signed meter values.
+ */
+export class SignedMeterValuesUtil {
+  private readonly _fileAccess: IFileAccess;
+  private readonly _logger: Logger<ILogObj>;
+  private readonly _deviceModelRepository: IDeviceModelRepository;
+  private readonly _chargingStationSecurityInfoRepository: IChargingStationSecurityInfoRepository;
+
+  private readonly _signedMeterValuesConfiguration:
+    | SignedMeterValuesConfig
+    | undefined;
+
+  /**
+   * @param {IFileAccess} [fileAccess] - The `fileAccess` allows access to the configured file storage.
+   *
+   * @param {SystemConfig} config - The `config` contains the current system configuration settings.
+   *
+   * @param {Logger<ILogObj>} [logger] - The `logger` represents an instance of {@link Logger<ILogObj>}.
+   *
+   * @param {IDeviceModelRepository} [deviceModelRepository] - The `deviceModelRepository` represents a repository for accessing and manipulating variable attribute data.
+   *
+   */
+  constructor(
+    fileAccess: IFileAccess,
+    config: SystemConfig,
+    logger: Logger<ILogObj>,
+    deviceModelRepository: IDeviceModelRepository,
+  ) {
+    this._fileAccess = fileAccess;
+    this._logger = logger;
+    this._deviceModelRepository = deviceModelRepository;
+    this._chargingStationSecurityInfoRepository =
+      new sequelize.SequelizeChargingStationSecurityInfoRepository(
+        config,
+        logger,
+      );
+
+    this._signedMeterValuesConfiguration =
+      config.modules.transactions.signedMeterValuesConfiguration;
+  }
+
+  public async validateMeterValues(
+    stationId: string,
+    meterValues: [MeterValueType, ...MeterValueType[]],
+  ): Promise<boolean> {
+    let validMeterValues = true;
+
+    const expectPublicKey =
+      await this._deviceModelRepository.readAllByQuerystring({
+        stationId,
+        component_name: 'OCPPCommCtrlr',
+        variable_name: 'PublicKeyWithSignedMeterValue',
+      });
+
+    for (const meterValue of meterValues) {
+      if (expectPublicKey.length > 0 && expectPublicKey[0].value !== 'Never') {
+        for (const sampledValue of meterValue.sampledValue) {
+          const signedMeterValue = sampledValue.signedMeterValue;
+
+          if (!signedMeterValue) {
+            continue;
+          }
+
+          if (
+            signedMeterValue.publicKey &&
+            signedMeterValue.publicKey.length > 0
+          ) {
```

Comment:
Isn't `publicKey` required if `signedMeterValue` is present?

Reply:
unfortunately no :( based on the section `2.7.17. PublicKeyWithSignedMeterValue` in the `OCPP-2.0.1_edition3_part2_specification.pdf`:

```
This Configuration Variable can be used to configure whether a public key needs to be sent with a signed meter
value. Note, that the field is required, so it needs to be present as an empty string when the public key is not sent.
```

Reply:
Gotcha. So if public key is not sent, use the public key stored on the CSMS? Which is presumably the same public key the charger used to encrypt the message?

Reply:
yup exactly, if the public key is not sent in the signed meter values message, find the public key stored in the csms. 

for right now, if the public key is in the signed meter values message, we go ahead and create the key if it doesn't exist, or just confirm it exists, but don't actually do anything with the key at that time. it's mostly just a sanity check 😅